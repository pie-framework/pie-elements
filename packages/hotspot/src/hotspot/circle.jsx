import React from 'react';
import PropTypes from 'prop-types';
import { Circle, Group } from 'react-konva';
import { withStyles } from '@material-ui/core/styles';
import { ImageComponent } from '@pie-lib/pie-toolbox/icons';
import { faCorrect, faWrong } from './icons';

class CircleComponent extends React.Component {
  constructor(props) {
    super(props);
  }

  handleClick = (e) => {
    const { onClick, id, selected, disabled } = this.props;

    if (!disabled) {
      e.cancelBubble = true;
      onClick({ id, selected: !selected });
    }
  };

  handleMouseEnter = () => {
    const { disabled } = this.props;

    if (!disabled) {
      document.body.style.cursor = 'pointer';
    }
  };

  handleMouseLeave = () => {
    document.body.style.cursor = 'default';
  };

  getEvaluateOutlineColor = (isCorrect, markAsCorrect, outlineColor) =>
    markAsCorrect ? 'green' : isCorrect ? outlineColor : 'red';

  getOutlineWidth = (showCorrectEnabled, selected, markAsCorrect, strokeWidth) =>
    markAsCorrect || (!markAsCorrect && !showCorrectEnabled && selected) ? strokeWidth : 0;

  render() {
    const {
      classes,
      radius,
      hotspotColor,
      isCorrect,
      isEvaluateMode,
      outlineColor,
      selected,
      x,
      y,
      evaluateText,
      strokeWidth,
      scale,
      markAsCorrect,
      showCorrectEnabled,
    } = this.props;

    const outlineColorParsed = isEvaluateMode
      ? this.getEvaluateOutlineColor(isCorrect, markAsCorrect, outlineColor)
      : outlineColor;

    const outlineWidth = this.getOutlineWidth(showCorrectEnabled, selected, markAsCorrect, strokeWidth);

    const iconX = x - 10;
    const iconY = y - 10; // Adjust position for the icon

    let iconSrc;
    if (showCorrectEnabled) {
      if ((selected && isCorrect) || (!selected && !isCorrect)) {
        iconSrc = faCorrect;
      }
    } else {
      if (selected) {
        if (isCorrect) {
          iconSrc = faCorrect;
        } else {
          iconSrc = faWrong;
        }
      } else if (!isCorrect) {
        iconSrc = faWrong;
      }
    }

    return (
      <Group scaleX={scale} scaleY={scale}>
        <Circle
          classes={classes.base}
          radius={radius}
          fill={hotspotColor}
          onClick={this.handleClick}
          onTap={this.handleClick}
          draggable={false}
          stroke={outlineColorParsed}
          strokeWidth={outlineWidth}
          onMouseLeave={this.handleMouseLeave}
          onMouseEnter={this.handleMouseEnter}
          x={x}
          y={y}
        />
        {isEvaluateMode && iconSrc ? <ImageComponent src={iconSrc} x={iconX} y={iconY} tooltip={evaluateText} /> : null}
      </Group>
    );
  }
}

const styles = () => ({
  base: {
    cursor: 'pointer',
    opacity: 0.5,
    position: 'relative',
  },
});

CircleComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  radius: PropTypes.number.isRequired,
  hotspotColor: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  isCorrect: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  isEvaluateMode: PropTypes.bool.isRequired,
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  outlineColor: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  evaluateText: PropTypes.string,
  strokeWidth: PropTypes.number,
  scale: PropTypes.number,
  markAsCorrect: PropTypes.bool.isRequired,
  showCorrectEnabled: PropTypes.bool.isRequired,
};

CircleComponent.defaultProps = {
  isCorrect: false,
  evaluateText: null,
  strokeWidth: 5,
  scale: 1,
};

export default withStyles(styles)(CircleComponent);
