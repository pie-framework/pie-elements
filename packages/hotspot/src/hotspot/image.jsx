import React from 'react';
import PropTypes from 'prop-types';
import { Image } from 'react-konva';

class ImageComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      image: null
    };
  }

  componentDidMount() {
    this.loadImage();
  }

  componentDidUpdate(oldProps) {
    if (oldProps.src !== this.props.src) {
      this.loadImage();
    }
  }

  componentWillUnmount() {
    this.image.removeEventListener('load', this.handleLoad);
  }

  loadImage() {
    const { src } = this.props;

    this.image = new window.Image();
    this.image.src = src;
    this.image.addEventListener('load', this.handleLoad);
  }

  handleLoad = () => {
    this.setState({
      image: this.image
    });
  };

  render() {
    const { x, y } = this.props;
    const { image } = this.state;

    return (
      <Image
        width={20}
        height={20}
        x={x}
        y={y}
        image={image}
      />
    );
  }
}

ImageComponent.propTypes = {
  src: PropTypes.string.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired
};

export default ImageComponent;
