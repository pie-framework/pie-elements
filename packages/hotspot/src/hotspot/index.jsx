import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import CorrectAnswerToggle from '@pie-lib/correct-answer-toggle';
import { Collapsible, hasText, PreviewPrompt } from '@pie-lib/render-ui';
import { withStyles } from '@material-ui/core/styles';

import Container from './container';

class HotspotComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showCorrect: false,
      observer: null,
      scale: 1,
    };
  }

  componentDidMount() {
    this.observer = new MutationObserver((mutations) => {
      mutations.forEach(() => {
        const target = document.getElementById('question-container')?.style?.cssText;
        const zoom = target?.substring(target.indexOf('--pie-zoom') + 11, target.lastIndexOf('%'));

        const zoomParsed = zoom?.replace(/\s/g, '');
        if (zoomParsed) {
          const newScale = parseFloat(zoomParsed) / 100;
          if (newScale !== this.state.scale) {
            this.setState({
              scale: parseFloat(zoomParsed) / 100,
            });
          }
        } else if (!zoomParsed && this.state.scale !== 1) {
          this.setState({
            scale: 1,
          });
        }
      });
    });

    const target = document.getElementById('question-container');
    if (target) {
      this.observer.observe(target, { attributes: true, attributeFilter: ['style'] });
    }
  }

  componentWillUnmount() {
    this.observer?.disconnect();
  }

  onToggle = () => {
    const { showCorrect } = this.state;
    this.setState({ showCorrect: !showCorrect });
  };

  render() {
    const {
      session,
      model: {
        disabled,
        imageUrl,
        prompt,
        mode,
        multipleCorrect,
        shapes,
        outlineColor,
        hotspotColor,
        dimensions,
        rationale,
        teacherInstructions,
        strokeWidth,
        responseCorrect,
      },
      onSelectChoice,
      classes,
    } = this.props;
    const { showCorrect } = this.state;
    const isEvaluateMode = mode === 'evaluate';
    const showCorrectAnswerToggle = isEvaluateMode && !responseCorrect;

    return (
      <div>
        {showCorrectAnswerToggle && (
          <CorrectAnswerToggle
            show={showCorrectAnswerToggle}
            toggled={showCorrect}
            onToggle={this.onToggle.bind(this)}
          />
        )}
        {showCorrectAnswerToggle && <br />}
        {teacherInstructions && hasText(teacherInstructions) && (
          <Collapsible
            labels={{ hidden: 'Show Teacher Instructions', visible: 'Hide Teacher Instructions' }}
            className={classes.collapsible}
          >
            <PreviewPrompt className="prompt" prompt={teacherInstructions} />
          </Collapsible>
        )}

        <Typography className={classes.prompt}>
          <PreviewPrompt className="prompt" prompt={prompt} />
        </Typography>

        {imageUrl ? (
          <Container
            isEvaluateMode={isEvaluateMode}
            session={session}
            dimensions={dimensions}
            imageUrl={imageUrl}
            hotspotColor={hotspotColor}
            multipleCorrect={multipleCorrect}
            outlineColor={outlineColor}
            onSelectChoice={onSelectChoice}
            shapes={shapes}
            disabled={disabled}
            strokeWidth={strokeWidth}
            scale={this.state.scale}
            showCorrect={showCorrect}
          />
        ) : null}

        {rationale && hasText(rationale) && (
          <Collapsible labels={{ hidden: 'Show Rationale', visible: 'Hide Rationale' }} className={classes.collapsible}>
            <PreviewPrompt className="prompt" prompt={rationale} />
          </Collapsible>
        )}
      </div>
    );
  }
}

HotspotComponent.propTypes = {
  classes: PropTypes.object,
  model: PropTypes.object.isRequired,
  onSelectChoice: PropTypes.func.isRequired,
  session: PropTypes.object.isRequired,
};

HotspotComponent.defaultProps = {
  classes: {},
};

const styles = (theme) => ({
  collapsible: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
  },
  prompt: {
    fontSize: 'inherit',
  },
});

export default withStyles(styles)(HotspotComponent);
