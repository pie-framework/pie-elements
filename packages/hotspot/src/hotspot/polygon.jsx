import React from 'react';
import PropTypes from 'prop-types';
import { Line, Group, Rect } from 'react-konva';
import { withStyles } from '@material-ui/core/styles';
import { ImageComponent } from '@pie-lib/pie-toolbox/icons';
import { faCorrect, faWrong } from './icons';

class PolygonComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      hovered: false,
    };
  }

  getPolygonCenter = (points) => {
    const x = points.map(({ x }) => x);
    const y = points.map(({ y }) => y);
    const minX = Math.min.apply(null, x);
    const maxX = Math.max.apply(null, x);
    const minY = Math.min.apply(null, y);
    const maxY = Math.max.apply(null, y);
    return [(minX + maxX) / 2, (minY + maxY) / 2];
  };

  parsePointsForKonva = (points) => {
    const parsedPoints = [];
    points.forEach(({ x, y }) => {
      parsedPoints.push(x);
      parsedPoints.push(y);
    });
    return parsedPoints;
  };

  handleClick = (e) => {
    const { onClick, id, selected, disabled } = this.props;

    if (!disabled) {
      e.cancelBubble = true;
      onClick({ id, selected: !selected, selector: 'Mouse' });
    }
  };

  handleMouseEnter = () => {
    const { disabled } = this.props;

    if (!disabled) {
      document.body.style.cursor = 'pointer';
    }
    this.setState({ hovered: true });
  };

  handleMouseLeave = () => {
    document.body.style.cursor = 'default';
    this.setState({ hovered: false });
  };

  getEvaluateOutlineColor = (isCorrect, markAsCorrect, outlineColor) =>
    markAsCorrect ? 'green' : isCorrect ? outlineColor : 'red';

  getOutlineWidth = (showCorrectEnabled, selected, markAsCorrect, strokeWidth) =>
    markAsCorrect || (!markAsCorrect && !showCorrectEnabled && selected) ? strokeWidth : 0;

  render() {
    const {
      classes,
      hotspotColor,
      isCorrect,
      isEvaluateMode,
      hoverOutlineColor,
      outlineColor,
      selected,
      points,
      evaluateText,
      strokeWidth,
      scale,
      markAsCorrect,
      selectedHotspotColor,
      showCorrectEnabled,
    } = this.props;

    const { hovered } = this.state;

    const outlineColorParsed = isEvaluateMode
      ? this.getEvaluateOutlineColor(isCorrect, markAsCorrect, outlineColor)
      : outlineColor;
    const outlineWidth = this.getOutlineWidth(showCorrectEnabled, selected, markAsCorrect, strokeWidth);

    const pointsParsed = this.parsePointsForKonva(points);
    const center = this.getPolygonCenter(points);
    const iconX = center[0];
    const iconY = center[1];

    // "Show Correct Answer" Enabled:
    //   - Correctly Selected: white checkmark in green circle
    //   - Correctly Not Selected: none
    //   - Incorrectly Selected: none
    //   - Incorrectly Not Selected: white checkmark in green circle
    // "Show Correct Answer" Disabled:
    //   - Correctly Selected:
    //     - white checkmark in green circle
    //     - heavy outline, as on “Gather”
    //   - Correctly Not Selected: none
    //   - Incorrectly Selected:
    //     - white "X" in red circle
    //     - heavy outline around the selection should appear in red
    //   - Incorrectly Not Selected: white "X" in red circle
    let iconSrc;

    if (showCorrectEnabled) {
      if ((selected && isCorrect) || (!selected && !isCorrect)) {
        iconSrc = faCorrect;
      }
    } else {
      if (selected) {
        if (isCorrect) {
          iconSrc = faCorrect;
        } else {
          iconSrc = faWrong;
        }
      } else if (!isCorrect) {
        iconSrc = faWrong;
      }
    }
    const useHoveredStyle = hovered && hoverOutlineColor;

    const xValues = pointsParsed.filter((_, index) => index % 2 === 0); // Even indices are x-coordinates
    const yValues = pointsParsed.filter((_, index) => index % 2 !== 0); // Odd indices are y-coordinates

    const minX = Math.min(...xValues);
    const maxX = Math.max(...xValues);
    const minY = Math.min(...yValues);
    const maxY = Math.max(...yValues);

    const rectX = minX;
    const rectY = minY;
    const rectWidth = maxX - minX;
    const rectHeight = maxY - minY;

    return (
      <Group scaleX={scale} scaleY={scale}>
        {useHoveredStyle && (
          <Rect
            x={rectX}
            y={rectY}
            width={rectWidth}
            height={rectHeight}
            stroke={selected ? 'transparent' : hoverOutlineColor}
            strokeWidth={strokeWidth}
          />
        )}
        <Line
          points={pointsParsed}
          closed={true}
          classes={classes.base}
          fill={selected && selectedHotspotColor? selectedHotspotColor : hotspotColor}
          onClick={this.handleClick}
          onTap={this.handleClick}
          draggable={false}
          stroke={useHoveredStyle && !selected ? 'transparent' : outlineColorParsed}
          strokeWidth={useHoveredStyle && !selected ? 0 : outlineWidth}
          onMouseLeave={this.handleMouseLeave}
          onMouseEnter={this.handleMouseEnter}
        />
        {isEvaluateMode && iconSrc ? <ImageComponent src={iconSrc} x={iconX} y={iconY} tooltip={evaluateText} /> : null}
      </Group>
    );
  }
}

const styles = () => ({
  base: {
    cursor: 'pointer',
    opacity: 0.5,
    position: 'relative',
  },
});

PolygonComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  hotspotColor: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  isCorrect: PropTypes.bool.isRequired,
  isEvaluateMode: PropTypes.bool.isRequired,
  hoverOutlineColor: PropTypes.string,
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  outlineColor: PropTypes.string.isRequired,
  points: PropTypes.array.isRequired,
  selected: PropTypes.bool.isRequired,
  evaluateText: PropTypes.string,
  selectedHotspotColor: PropTypes.string,
  strokeWidth: PropTypes.number,
  scale: PropTypes.number,
  markAsCorrect: PropTypes.bool.isRequired,
  showCorrectEnabled: PropTypes.bool.isRequired,
};

PolygonComponent.defaultProps = {
  evaluateText: null,
  strokeWidth: 5,
  scale: 1,
};

export default withStyles(styles)(PolygonComponent);
