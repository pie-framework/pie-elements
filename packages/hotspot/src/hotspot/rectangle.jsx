import React from 'react';
import PropTypes from 'prop-types';
import { Rect, Group } from 'react-konva';
import { withStyles } from '@material-ui/core/styles';
import ImageComponent from './image-konva-tooltip';
import { faCorrect, faWrong } from './icons';

class RectComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      hovered: false,
    };
  }

  handleClick = (e) => {
    const { onClick, id, selected, disabled } = this.props;

    if (!disabled) {
      e.cancelBubble = true;
      onClick({ id, selected: !selected, selector: 'Mouse' });
    }
  };

  handleMouseEnter = () => {
    const { disabled } = this.props;

    if (!disabled) {
      document.body.style.cursor = 'pointer';
    }
    this.setState({ hovered: true });
  };

  handleMouseLeave = () => {
    document.body.style.cursor = 'default';
    this.setState({ hovered: false });
  };

  getEvaluateOutlineColor = (isCorrect, markAsCorrect, outlineColor) =>
    markAsCorrect ? 'green' : isCorrect ? outlineColor : 'red';

  getOutlineWidth = (showCorrectEnabled, selected, markAsCorrect, strokeWidth) =>
    markAsCorrect || (!markAsCorrect && !showCorrectEnabled && selected) ? strokeWidth : 0;

  render() {
    const {
      classes,
      height,
      hotspotColor,
      hoverOutlineColor,
      selectedHotspotColor,
      isCorrect,
      isEvaluateMode,
      outlineColor,
      selected,
      width,
      x,
      y,
      evaluateText,
      strokeWidth,
      scale,
      markAsCorrect,
      showCorrectEnabled,
    } = this.props;

    const outlineColorParsed = isEvaluateMode
      ? this.getEvaluateOutlineColor(isCorrect, markAsCorrect, outlineColor)
      : outlineColor;

    const outlineWidth = this.getOutlineWidth(showCorrectEnabled, selected, markAsCorrect, strokeWidth);

    const iconX = x + width / 2 - 10;
    const iconY = y + height / 2 - 10;

    // "Show Correct Answer" Enabled:
    //   - Correctly Selected: white checkmark in green circle
    //   - Correctly Not Selected: none
    //   - Incorrectly Selected: none
    //   - Incorrectly Not Selected: white checkmark in green circle
    // "Show Correct Answer" Disabled:
    //   - Correctly Selected:
    //     - white checkmark in green circle
    //     - heavy outline, as on “Gather”
    //   - Correctly Not Selected: none
    //   - Incorrectly Selected:
    //     - white "X" in red circle
    //     - heavy outline around the selection should appear in red
    //   - Incorrectly Not Selected: white "X" in red circle
    let iconSrc;

    if (showCorrectEnabled) {
      if ((selected && isCorrect) || (!selected && !isCorrect)) {
        iconSrc = faCorrect;
      }
    } else {
      if (selected) {
        if (isCorrect) {
          iconSrc = faCorrect;
        } else {
          iconSrc = faWrong;
        }
      } else if (!isCorrect) {
        iconSrc = faWrong;
      }
    }

    const { hovered } = this.state;
    const useHoveredStyle = hovered && hoverOutlineColor;

    return (
      <Group scaleX={scale} scaleY={scale}>
        {useHoveredStyle && (
          <Rect
            x={x}
            y={y}
            width={width}
            height={height}
            stroke={selected ? 'transparent' : hoverOutlineColor}
            strokeWidth={strokeWidth}
            listening={false}
          />
        )}
        <Rect
          classes={classes.base}
          width={width}
          height={height}
          fill={selected && selectedHotspotColor ? selectedHotspotColor : hotspotColor}
          onClick={this.handleClick}
          onTap={this.handleClick}
          draggable={false}
          stroke={useHoveredStyle && !selected ? 'transparent' : outlineColorParsed}
          strokeWidth={useHoveredStyle && !selected ? 0 : outlineWidth}
          onMouseLeave={this.handleMouseLeave}
          onMouseEnter={this.handleMouseEnter}
          x={x}
          y={y}
        />
        {isEvaluateMode && iconSrc ? <ImageComponent src={iconSrc} x={iconX} y={iconY} tooltip={evaluateText} /> : null}
      </Group>
    );
  }
}

const styles = () => ({
  base: {
    cursor: 'pointer',
    opacity: 0.5,
    position: 'relative',
  },
});

RectComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  height: PropTypes.number.isRequired,
  hotspotColor: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  isCorrect: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  isEvaluateMode: PropTypes.bool.isRequired,
  hoverOutlineColor: PropTypes.string,
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  outlineColor: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired,
  width: PropTypes.number.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  evaluateText: PropTypes.string,
  strokeWidth: PropTypes.number,
  scale: PropTypes.number,
  selectedHotspotColor: PropTypes.string,
  markAsCorrect: PropTypes.bool.isRequired,
  showCorrectEnabled: PropTypes.bool.isRequired,
};

RectComponent.defaultProps = {
  isCorrect: false,
  evaluateText: null,
  strokeWidth: 5,
  scale: 1,
};

export default withStyles(styles)(RectComponent);
