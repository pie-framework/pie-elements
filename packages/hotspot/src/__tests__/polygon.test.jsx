import React from 'react';
import _ from 'lodash';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Konva from 'konva';
import { Line } from 'react-konva';

import { shallowChild } from '@pie-lib/test-utils';

import Polygon from '../hotspot/polygon';
import Image from '../hotspot/image';
import { faCorrect, faWrong } from '../hotspot/icons';

Konva.isBrowser = false;

describe('Polygon', () => {
  let onClick, wrapper;

  const mkWrapper = (opts = {}) => {
    opts = _.extend(
      {
        classes: {
          base: 'base',
        },
        height: 200,
        hotspotColor: 'rgba(137, 183, 244, 0.65)',
        id: '1',
        isCorrect: false,
        isEvaluateMode: false,
        evaluateText: null,
        disabled: false,
        outlineColor: 'blue',
        selected: false,
        points: [
          { x: 94, y: 4 },
          { x: 89, y: 4 },
          { x: 36, y: 40 },
        ],
      },
      opts,
    );

    return shallow(<Polygon {...opts} onClick={onClick} />);
  };

  beforeEach(() => {
    onClick = jest.fn();
    wrapper = mkWrapper();
  });

  describe('snapshots', () => {
    describe('outline color', () => {
      it('renders', () => {
        const wrapper = mkWrapper({ outlineColor: 'red' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('outline width', () => {
      it('renders with default border width', () => {
        const wrapper = mkWrapper();
        expect(toJson(wrapper)).toMatchSnapshot();
      });

      it('renders with given border width', () => {
        const wrapper = mkWrapper({ strokeWidth: 10 });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('hotspot color', () => {
      it('renders', () => {
        const wrapper = mkWrapper({ hotspotColor: 'rgba(217, 30, 24, 0.65)' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('evaluate with correct answer', () => {
      it('renders', () => {
        const wrapper = mkWrapper({ isEvaluateMode: true, isCorrect: true, evaluateText: 'Correctly\nselected' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe('in evaluate mode', () => {
    const defaultModel = {
      height: 200,
      hotspotColor: 'rgba(137, 183, 244, 0.65)',
      id: '1',
      isCorrect: false,
      isEvaluateMode: true,
      evaluateText: null,
      disabled: true,
      outlineColor: 'blue',
      strokeWidth: 5,
      selected: false,
      points: [
        { x: 94, y: 4 },
        { x: 89, y: 4 },
        { x: 36, y: 40 },
      ],
      markAsCorrect: false,
      showCorrectEnabled: false,
      onClick,
    };

    const getComponents = (model) => {
      const testWrapper = shallowChild(
        Polygon,
        {
          ...defaultModel,
          ...model,
        },
        1,
      );
      const polygonComponent = testWrapper();
      const lineComponent = polygonComponent.find(Line);
      return {
        polygonComponent,
        lineComponent,
      };
    };

    describe('when correctly selected', () => {
      const { polygonComponent, lineComponent } = getComponents({
        selected: true,
        isCorrect: true,
      });

      it('should have a blue outline', () => {
        expect(lineComponent.prop('stroke')).toEqual('blue');
        expect(lineComponent.prop('strokeWidth')).toEqual(defaultModel.strokeWidth);
      });

      it('should have a green checkmark icon', () => {
        const imgComponent = polygonComponent.find(Image);
        expect(imgComponent.length).toEqual(1);
        expect(imgComponent.prop('src')).toEqual(faCorrect);
      });
    });

    describe('when correctly not selected', () => {
      const { polygonComponent, lineComponent } = getComponents({
        selected: false,
        isCorrect: true,
      });

      it('should have no outline', () => {
        expect(lineComponent.prop('strokeWidth')).toEqual(0);
      });

      it('should have no icon', () => {
        const imgComponent = polygonComponent.find(Image);
        expect(imgComponent.length).toEqual(0);
      });
    });

    describe('when incorrectly selected', () => {
      const { polygonComponent, lineComponent } = getComponents({
        selected: true,
        isCorrect: false,
      });

      it('should have a red outline', () => {
        expect(lineComponent.prop('stroke')).toEqual('red');
        expect(lineComponent.prop('strokeWidth')).toEqual(defaultModel.strokeWidth);
      });

      it('should have a red x icon', () => {
        const imgComponent = polygonComponent.find(Image);
        expect(imgComponent.length).toEqual(1);
        expect(imgComponent.prop('src')).toEqual(faWrong);
      });
    });

    describe('when incorrectly not selected', () => {
      const { polygonComponent, lineComponent } = getComponents({
        selected: false,
        isCorrect: false,
      });

      it('should have no outline', () => {
        expect(lineComponent.prop('strokeWidth')).toEqual(0);
      });

      it('should have a red x icon', () => {
        const imgComponent = polygonComponent.find(Image);
        expect(imgComponent.length).toEqual(1);
        expect(imgComponent.prop('src')).toEqual(faWrong);
      });
    });

    describe('when showing correct answer (showCorrectEnabled = true)', () => {
      it('should have a green outline', () => {
        const { lineComponent } = getComponents({
          showCorrectEnabled: true,
          markAsCorrect: true,
        });
        expect(lineComponent.prop('stroke')).toEqual('green');
        expect(lineComponent.prop('strokeWidth')).toEqual(defaultModel.strokeWidth);
      });

      it('should have a green checkmark icon', () => {
        const { polygonComponent } = getComponents({
          showCorrectEnabled: true,
          markAsCorrect: true,
        });
        const imgComponent = polygonComponent.find(Image);
        expect(imgComponent.prop('src')).toEqual(faCorrect);
      });

      it('should not be selected if the answer is incorrect', () => {
        const { lineComponent } = getComponents({
          showCorrectEnabled: true,
          markAsCorrect: false,
          selected: true,
          isCorrect: false,
        });
        expect(lineComponent.prop('strokeWidth')).toEqual(0);
      });

      it('should not render an icon if the answer is incorrect', () => {
        const { polygonComponent } = getComponents({
          showCorrectEnabled: true,
          markAsCorrect: false,
          selected: true,
          isCorrect: false,
        });
        const imgComponent = polygonComponent.find(Image);
        expect(imgComponent.length).toEqual(0);
      });
    });
  });
});
