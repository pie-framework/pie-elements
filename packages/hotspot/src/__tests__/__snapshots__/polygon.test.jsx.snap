// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polygon snapshots evaluate with correct answer renders 1`] = `
<PolygonComponent
  classes={
    Object {
      "base": "PolygonComponent-base-1 base",
    }
  }
  disabled={false}
  evaluateText="Correctly
selected"
  height={200}
  hotspotColor="rgba(137, 183, 244, 0.65)"
  id="1"
  isCorrect={true}
  isEvaluateMode={true}
  onClick={[MockFunction]}
  outlineColor="blue"
  points={
    Array [
      Object {
        "x": 94,
        "y": 4,
      },
      Object {
        "x": 89,
        "y": 4,
      },
      Object {
        "x": 36,
        "y": 40,
      },
    ]
  }
  scale={1}
  selected={false}
  strokeWidth={5}
/>
`;

exports[`Polygon snapshots hotspot color renders 1`] = `
<PolygonComponent
  classes={
    Object {
      "base": "PolygonComponent-base-1 base",
    }
  }
  disabled={false}
  evaluateText={null}
  height={200}
  hotspotColor="rgba(217, 30, 24, 0.65)"
  id="1"
  isCorrect={false}
  isEvaluateMode={false}
  onClick={[MockFunction]}
  outlineColor="blue"
  points={
    Array [
      Object {
        "x": 94,
        "y": 4,
      },
      Object {
        "x": 89,
        "y": 4,
      },
      Object {
        "x": 36,
        "y": 40,
      },
    ]
  }
  scale={1}
  selected={false}
  strokeWidth={5}
/>
`;

exports[`Polygon snapshots outline color renders 1`] = `
<PolygonComponent
  classes={
    Object {
      "base": "PolygonComponent-base-1 base",
    }
  }
  disabled={false}
  evaluateText={null}
  height={200}
  hotspotColor="rgba(137, 183, 244, 0.65)"
  id="1"
  isCorrect={false}
  isEvaluateMode={false}
  onClick={[MockFunction]}
  outlineColor="red"
  points={
    Array [
      Object {
        "x": 94,
        "y": 4,
      },
      Object {
        "x": 89,
        "y": 4,
      },
      Object {
        "x": 36,
        "y": 40,
      },
    ]
  }
  scale={1}
  selected={false}
  strokeWidth={5}
/>
`;

exports[`Polygon snapshots outline width renders with default border width 1`] = `
<PolygonComponent
  classes={
    Object {
      "base": "PolygonComponent-base-1 base",
    }
  }
  disabled={false}
  evaluateText={null}
  height={200}
  hotspotColor="rgba(137, 183, 244, 0.65)"
  id="1"
  isCorrect={false}
  isEvaluateMode={false}
  onClick={[MockFunction]}
  outlineColor="blue"
  points={
    Array [
      Object {
        "x": 94,
        "y": 4,
      },
      Object {
        "x": 89,
        "y": 4,
      },
      Object {
        "x": 36,
        "y": 40,
      },
    ]
  }
  scale={1}
  selected={false}
  strokeWidth={5}
/>
`;

exports[`Polygon snapshots outline width renders with given border width 1`] = `
<PolygonComponent
  classes={
    Object {
      "base": "PolygonComponent-base-1 base",
    }
  }
  disabled={false}
  evaluateText={null}
  height={200}
  hotspotColor="rgba(137, 183, 244, 0.65)"
  id="1"
  isCorrect={false}
  isEvaluateMode={false}
  onClick={[MockFunction]}
  outlineColor="blue"
  points={
    Array [
      Object {
        "x": 94,
        "y": 4,
      },
      Object {
        "x": 89,
        "y": 4,
      },
      Object {
        "x": 36,
        "y": 40,
      },
    ]
  }
  scale={1}
  selected={false}
  strokeWidth={10}
/>
`;
